modules/packages/PrecisionSerializer.chpl

```diff
@@ -0,0 +1,575 @@
+                    the writer used to create current AggregateSerializer.
+        :arg name: The name of the class type.
+        :arg size: The number of fields in the class.
+
+        :returns: A new :record:`~PrecisionSerializer.precisionSerializer.AggregateSerializer`
+      */
+      proc ref startClass(writer: fileWriter, name: string, size: int) throws {
+        // Note: 'size' of parent might be zero, but 'size' of grandparent might
+        // be non-zero.
+        return new AggregateSerializer(writer, _parent=true,
+                                      _firstPtr=c_addrOf(_first));
+      }
+
+      /*
+        Ends serialization of the current class by writing the character ``}``
+
+        .. note:: It is an error to call methods on an AggregateSerializer after
+                  invoking 'endClass'.
+      */
+      proc endClass() throws {
+        if !_parent then
+          writer.writeLiteral(_ending);
+        else if _firstPtr != nil then
+          _firstPtr.deref() = _first;
+      }
+
+      /*
+        Ends serialization of the current record by writing the character ``)``
+
+        .. note:: It is an error to call methods on an AggregateSerializer after
+                  invoking 'endRecord'.
+      */
+      proc endRecord() throws {
+        writer.writeLiteral(_ending);
+      }
+    }
+
+    /*
+      Start serializing a tuple by writing the character ``(``.
+
+      :arg writer: The :record:`~IO.fileWriter` to be used when serializing.
+      :arg size: The number of elements in the tuple.
+
+      :returns: A new :record:`TupleSerializer`
+    */
+    proc startTuple(writer: fileWriter, size: int) throws {
+      writer.writeLiteral("(");
+      return new TupleSerializer(writer, size);
+    }
+
+    /*
+      Returned by ``startTuple`` to provide the API for serializing tuples.
+
+      A tuple will be serialized as a comma-separated list between two
+      parentheses. For example, the tuple literal ``(1, 2, 3)`` would be
+      serialized as:
+
+      .. code-block::
+
+        (1, 2, 3)
+
+      A 1-tuple will be serialized with a trailing comma. For example, the literal
+      ``(4,)`` would be serialized as ``(4,)``.
+
+    */
+    record TupleSerializer {
+      @chpldoc.nodoc
+      var writer;
+      @chpldoc.nodoc
+      const size : int;
+      @chpldoc.nodoc
+      var _first : bool = true;
+
+      /*
+        Serialize ``element``.
+
+        Writes a leading comma before serializing the element if this is not the
+        first element in the tuple.
+      */
+      proc ref writeElement(const element: ?) throws {
+        if !_first then writer.writeLiteral(", ");
+        else _first = false;
+
+        writer.write(element);
+      }
+
+      /*
+        Ends serialization of the current tuple by writing the character ``)``.
+
+        Adds a comma between the last value and ``)`` if there was only one
+        element.
```

Comment:
Should there also be an error note here about calling other methods on the TupleSerializer after "endTuple"?